File: app/Sources/Myco/AppStore.swift
import SwiftUI
import Combine

/// 全局状态中枢：主题、tab、检测到的 agent、可分发 skill、会话列表、运行日志。
final class AppStore: ObservableObject {
    @Published var appearance: Appearance = .dark
    @Published var tab: Tab = .home
    @Published var agents: [Agent] = []
    @Published var skills: [ShareableSkill] = []
    @Published var sessions: [Session] = []
    @Published var busy = false
    @Published var lastLog: String = ""
    @Published var sessionsLoading = false   // 真实会话异步加载中

    var palette: Palette { Palette(appearance: appearance) }

    enum Tab: String, CaseIterable { case home, share, relay, history, settings }

    // 检测到并安装的 agent
    var installed: [Agent] { agents.filter { $0.installed } }
    var totalSessions: Int { agents.reduce(0) { $0 + $1.sessionCount } }
    var installedCount: Int { installed.count }

    func toggleTheme() {
        withAnimation(.springSoft) {
            appearance = (appearance == .dark ? .light : .dark)
        }
    }

    /// 异步检测本机 agent + 加载可分发 skill；随后再异步拉取真实会话列表。
    /// 两段式：先秒出 agent 概览（快），会话列表（要扫全部历史，慢）后到。
    func refresh() {
        DispatchQueue.global(qos: .userInitiated).async {
            let detected = AgentDetector.detectAll()
            let skills = Self.loadShareableSkills()
            DispatchQueue.main.async {
                self.agents = detected
                self.skills = skills
                self.sessions = []
                self.loadRealSessions()   // 接真实会话（异步）
            }
        }
    }

    /// 调 handoff_chat.py --list --json 拉取真实会话，替换早期的占位数据。
    func loadRealSessions() {
        let installedIDs = installed.map { $0.id.rawValue }
        guard !installedIDs.isEmpty else { sessions = []; return }
        sessionsLoading = true
        PythonBridge.shared.handoffListSessions(agents: installedIDs) { [weak self] real in
            guard let self else { return }
            self.sessions = real
            self.sessionsLoading = false
        }
    }

    /// 从随附的 skills/ 目录读取可分发的 SKILL.md（真实存在的）。
    static func loadShareableSkills() -> [ShareableSkill] {
        let root = PythonBridge.shared.resourceRoot.appendingPathComponent("skills")
        var out: [ShareableSkill] = []
        let fm = FileManager.default
        if let items = try? fm.contentsOfDirectory(at: root, includingPropertiesForKeys: nil) {
            for dir in items where (try? dir.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true {
                let md = dir.appendingPathComponent("SKILL.md")
                guard fm.fileExists(atPath: md.path) else { continue }
                let text = (try? String(contentsOf: md, encoding: .utf8)) ?? ""
                let name = frontmatter(text, "name") ?? dir.lastPathComponent
                let desc = frontmatter(text, "description") ?? "（无描述）"
                out.append(ShareableSkill(id: dir.lastPathComponent, name: name,
                                          desc: String(desc.prefix(90)),
                                          path: dir.path))
            }
        }
        if out.isEmpty {
            out = [ShareableSkill(id: "multi-agent-skill-sharing",
                                  name: "multi-agent-skill-sharing",
                                  desc: "把一份 skill 扇出到每个 agent 的仓库目录。",
                                  path: root.appendingPathComponent("multi-agent-skill-sharing").path)]
        }
        return out
    }

    private static func frontmatter(_ text: String, _ key: String) -> String? {
        for line in text.split(separator: "\n").prefix(20) {
            let l = line.trimmingCharacters(in: .whitespaces)
            if l.hasPrefix("\(key):") {
                return l.dropFirst(key.count + 1).trimmingCharacters(in: .whitespaces)
                    .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
            }
        }
        return nil
    }

    /// skill 分发目标：从注册表(agents.json)驱动，与检测端 5 个 agent 完全对齐，
    /// 再加上跨 agent 通用目录(.agents)。彻底消除"检测到的 ≠ 能分享的"。
    func defaultTargets() -> [SkillTarget] {
        let url = PythonBridge.shared.resourceRoot
            .appendingPathComponent("engine/agents.json")
        guard let data = try? Data(contentsOf: url),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return fallbackTargets()
        }
        let defaults = Set((obj["defaultDistribute"] as? [String]) ?? ["claude", "codex", "workbuddy", "agents"])
        var out: [SkillTarget] = []
        // 5 个 agent（来自 agents[]）
        if let arr = obj["agents"] as? [[String: Any]] {
            for a in arr {
                guard let id = a["id"] as? String,
                      let dir = a["skillDir"] as? String else { continue }
                let label = (a["display"] as? String) ?? id
                out.append(SkillTarget(id: id, dir: dir, label: label,
                                       recommended: false, checked: defaults.contains(id)))
            }
        }
        // 额外跨 agent 目标（.agents / .cline）
        if let extra = obj["extraSkillTargets"] as? [[String: Any]] {
            for e in extra {
                guard let id = e["id"] as? String,
                      let dir = e["dir"] as? String else { continue }
                let label = (e["label"] as? String) ?? id
                out.append(SkillTarget(id: id, dir: dir, label: label,
                                       recommended: id == "agents", checked: defaults.contains(id)))
            }
        }
        return out.isEmpty ? fallbackTargets() : out
    }

    private func fallbackTargets() -> [SkillTarget] {
        [
            SkillTarget(id: "claude", dir: ".claude/skills", label: "Claude Code", recommended: false, checked: true),
            SkillTarget(id: "codex", dir: ".codex/skills", label: "Codex CLI", recommended: false, checked: true),
            SkillTarget(id: "workbuddy", dir: ".workbuddy/skills", label: "WorkBuddy", recommended: false, checked: true),
            SkillTarget(id: "agents", dir: ".agents/skills", label: "跨 agent 通用", recommended: true, checked: true),
            SkillTarget(id: "cline", dir: ".cline/skills", label: "Cline", recommended: false, checked: false),
        ]
    }
}

File: app-windows/AppStore.cs
using System.IO;
using System.Text.Json;

namespace Myco;

/// 全局状态中枢（对齐 macOS 版 AppStore.swift）：tab、agent、skill、会话、加载态。
/// 视图监听 StateChanged / TabChanged 重建。
public sealed class AppStore
{
    public static readonly AppStore I = new();

    public List<AgentVM> Agents { get; private set; } = new();
    public List<SkillVM> Skills { get; private set; } = new();
    public List<SessionVM> Sessions { get; private set; } = new();
    public bool SessionsLoading { get; private set; }
    public string Tab { get; private set; } = "home";

    /// agents/skills/sessions 任一批量变化后触发，视图整体重建。
    public event Action? StateChanged;
    public event Action<string>? TabChanged;

    public IEnumerable<AgentVM> Installed => Agents.Where(a => a.Installed);
    public int InstalledCount => Installed.Count();
    public int TotalSessions => Agents.Sum(a => a.Sessions);

    public void SetTab(string tab)
    {
        Tab = tab;
        TabChanged?.Invoke(tab);
    }

    /// 两段式刷新：先秒出 agent 概览（快），真实会话列表（扫全部历史，慢）后到。
    public async Task RefreshAsync()
    {
        var status = await PythonBridge.Shared.AgentStatusAsync();
        Agents = ParseAgents(status.Stdout);
        Skills = LoadShareableSkills();
        Sessions = new List<SessionVM>();
        StateChanged?.Invoke();

        var ids = Installed.Select(a => a.Id).ToList();
        if (ids.Count == 0) return;
        SessionsLoading = true;
        StateChanged?.Invoke();
        Sessions = await PythonBridge.Shared.HandoffListAsync(ids);
        SessionsLoading = false;
        StateChanged?.Invoke();
    }

    private static List<AgentVM> ParseAgents(string stdout)
    {
        var list = new List<AgentVM>();
        try
        {
            using var doc = JsonDocument.Parse(stdout);
            if (!doc.RootElement.TryGetProperty("agents", out var arr)) return list;
            foreach (var a in arr.EnumerateArray())
            {
                string S(string key, string fb = "") =>
                    a.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString()! : fb;
                bool Bl(string key) =>
                    a.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.True;
                var id = S("id");
                if (id.Length == 0) continue;
                list.Add(new AgentVM(
                    id, S("display", id), S("initial", id[..1].ToUpperInvariant()),
                    Bl("installed"),
                    a.TryGetProperty("sessions", out var n) && n.TryGetInt32(out var ni) ? ni : 0,
                    Bl("approximate"), Bl("pathChanged"), S("detail")));
            }
        }
        catch (JsonException) { }
        return list;
    }

    /// 从随附 skills/ 目录读取可分发的 SKILL.md（真实存在的）。
    private static List<SkillVM> LoadShareableSkills()
    {
        var root = Path.Combine(PythonBridge.Shared.ResourceRoot, "skills");
        var outList = new List<SkillVM>();
        if (Directory.Exists(root))
        {
            foreach (var dir in Directory.EnumerateDirectories(root))
            {
                var md = Path.Combine(dir, "SKILL.md");
                if (!File.Exists(md)) continue;
                var text = File.ReadAllText(md);
                var name = Frontmatter(text, "name") ?? Path.GetFileName(dir);
                var desc = Frontmatter(text, "description") ?? "（无描述）";
                outList.Add(new SkillVM(Path.GetFileName(dir), name,
                    desc.Length > 90 ? desc[..90] : desc, dir));
            }
        }
        return outList;
    }

    private static string? Frontmatter(string text, string key)
    {
        var lines = text.Split('\n').Take(20).ToArray();
        for (var i = 0; i < lines.Length; i++)
        {
            var l = lines[i].Trim();
            if (!l.StartsWith(key + ":")) continue;
            var val = l[(key.Length + 1)..].Trim().Trim('"', '\'');
            // YAML 块标量（>- / > / |- / |）：取后续缩进行拼成一段
            if (val is ">" or ">-" or "|" or "|-" or "")
            {
                var parts = new List<string>();
                for (var j = i + 1; j < lines.Length; j++)
                {
                    if (lines[j].Length == 0 || !char.IsWhiteSpace(lines[j][0])) break;
                    parts.Add(lines[j].Trim());
                }
                val = string.Join(" ", parts);
            }
            return val.Length > 0 ? val : null;
        }
        return null;
    }

    /// skill 分发目标：agents.json 驱动（agents[] + extraSkillTargets），
    /// defaultDistribute 决定默认勾选。
    public List<TargetVM> DefaultTargets()
    {
        var path = Path.Combine(PythonBridge.Shared.ResourceRoot, "engine", "agents.json");
        var outList = new List<TargetVM>();
        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            var root = doc.RootElement;
            var defaults = new HashSet<string>();
            if (root.TryGetProperty("defaultDistribute", out var dd))
                foreach (var d in dd.EnumerateArray())
                    if (d.GetString() is { } s) defaults.Add(s);

            if (root.TryGetProperty("agents", out var agents))
                foreach (var a in agents.EnumerateArray())
                {
                    var id = a.TryGetProperty("id", out var v) ? v.GetString() : null;
                    var dir = a.TryGetProperty("skillDir", out var v2) ? v2.GetString() : null;
                    if (id is null || dir is null) continue;
                    outList.Add(new TargetVM
                    {
                        Id = id, Dir = dir,
                        Label = a.TryGetProperty("display", out var v3) ? (v3.GetString() ?? id) : id,
                        Recommended = false, Checked = defaults.Contains(id),
                    });
                }
            if (root.TryGetProperty("extraSkillTargets", out var extra))
                foreach (var e in extra.EnumerateArray())
                {
                    var id = e.TryGetProperty("id", out var v) ? v.GetString() : null;
                    var dir = e.TryGetProperty("dir", out var v2) ? v2.GetString() : null;
                    if (id is null || dir is null) continue;
                    outList.Add(new TargetVM
                    {
                        Id = id, Dir = dir,
                        Label = e.TryGetProperty("label", out var v3) ? (v3.GetString() ?? id) : id,
                        Recommended = id == "agents", Checked = defaults.Contains(id),
                    });
                }
        }
        catch (Exception) { /* 注册表缺失/损坏时返回已收集到的部分 */ }
        return outList;
    }
}
